### DISCUSSION TITLE
How to gracefully handle API that sometimes return html instead of JSON

### DISCUSSION THREAD

**Discussion by mbajur** [lostisland/faraday#1223] [Help] [answered]
Hello there!

I have a JSON API client written using Faraday and from time to time, when something goes wrong internally on the API side, they return 50x error as a HTML error page instead of returning it as JSON. As i have a `conn.response :json, content_type: /\bjson$/` setting in place, Faraday tries to parse whatever API gives him as a JSON which result in json parsing exception (as it receives HTML). 

Is there any elegant way of dealing with such scenario? I could catch the exception outside of the client lib but i would rather stick to Faraday way of returning things, together with `success?` and `failure?` methods etc. 

For this specific case, i could even live without parsing the responses at all when 50x comes in.

Thanks in advance for any ideas.

**Comment by iMacTia** [comment, marked as answer]
Hi @mbajur, that's a bit strange as the `content_type: /\bjson$/` option exists specifically to solve this issue. That option in fact is a pattern match on the `Content-Type` header that triggers the response parsing if and only if the pattern matches the header (see: https://github.com/lostisland/faraday_middleware/blob/895301d88c5006cdf5b66d33b1c106863a5525bc/lib/faraday_middleware/response_middleware.rb#L37).

By the look of it, it seems as if the server is returning you a Content-Type: application/json header, while the body is in HTML.
That would of course be an issue on the server and would need to be resolved there.

Would it be possible for you to double-check the response headers of the server when status is 500 and body is HTML?

**Reply by mbajur** [reply]
Hey @iMacTia , i double checked and server returns Content-Type: application/json while response is a HTML. It's 100% server fault but i can't really do anything about it as it's a third-party server i have no control of.

**Reply by iMacTia** [reply]
Given the problem is only with 500 responses, I’d suggest you to use the [RaiseError](lib/faraday/response/raise_error.rb) middleware. Just remember to add it to your connection AFTER the json one (just before the adapter might be a good place).
This middleware will check if the response is 2xx and if not, it will raise an exception for you and interrupt the middleware chain 👍 

How does that sound? Let me know if you give it a try!

**Reply by mbajur** [reply]
Yes, i think that's my way to go. Thank you!
